browser/net/brave_ad_block_tp_network_delegate_helper.cc

```diff
@@ -215,26 +222,56 @@ class AdblockCnameResolveHostClient : public network::mojom::ResolveHostClient {
   }
 };
 
+// Called after checking a request without any CNAME uncloaking.
+void OnShouldBlockRequestResult(
+    scoped_refptr<base::SequencedTaskRunner> task_runner,
+    const ResponseCallback& next_callback,
+    std::shared_ptr<BraveRequestInfo> ctx) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  if (ctx->blocked_by == kAdBlocked) {
+    brave_shields::DispatchBlockedEvent(
+        ctx->request_url, ctx->render_frame_id, ctx->render_process_id,
+        ctx->frame_tree_node_id, brave_shields::kAds);
+    next_callback.Run();
+  } else {
+    // If not blocked, the request still needs to be CNAME uncloaked.
+
+    // DoH or standard DNS queries won't be routed through Tor, so we need to
+    // skip it.
+    if (ctx->browser_context->IsTor() || ctx->did_match_important) {
+      next_callback.Run();
+    } else {
+      // This will be deleted by `AdblockCnameResolveHostClient::OnComplete`.
+      new AdblockCnameResolveHostClient(std::move(next_callback), task_runner,
+                                        ctx);
+    }
+  }
+}
+
+void ShouldBlockRequest(scoped_refptr<base::SequencedTaskRunner> task_runner,
+                        const ResponseCallback& next_callback,
+                        std::shared_ptr<BraveRequestInfo> ctx) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  task_runner->PostTaskAndReply(
```

Comment:
it is possible to shorten such calls with `Then()` https://chromium.googlesource.com/chromium/src/+/master/docs/callback.md#chaining-callbacks

Reply:
Can you elaborate? It looks like that wouldn't apply here because we need to dispatch to the task runner and then back to the the UI thread

Reply:
sorry, I should have post a link to the next section :D https://chromium.googlesource.com/chromium/src/+/master/docs/callback.md#chaining-callbacks-across-different-task-runners

Reply:
hmm, if I understand correctly the equivalent would be:

```cpp
  task_runner->PostTask(
      FROM_HERE,
      base::BindOnce(&ShouldBlockRequestOnTaskRunner, ctx, base::nullopt)
          .Then(base::BindPostTask(
              content::GetUIThreadTaskRunner({}),
              base::BindOnce(&OnShouldBlockRequestResult, task_runner,
                             next_callback, ctx))));
```

which actually seems longer than the current version:

```cpp
  task_runner->PostTaskAndReply(
      FROM_HERE,
      base::BindOnce(&ShouldBlockRequestOnTaskRunner, ctx, base::nullopt),
      base::BindOnce(&OnShouldBlockRequestResult, task_runner, next_callback,
                     ctx));
```